File: src/common/Pizza.cs
using LanguageExt;

namespace common;

public abstract record PizzaToppingKind
{
    public sealed record Cheese : PizzaToppingKind
    {
        public static Cheese Instance { get; } = new();

        public override string ToString() => nameof(Cheese);
    }

    public sealed record Pepperoni : PizzaToppingKind
    {
        public static Pepperoni Instance { get; } = new();

        public override string ToString() => nameof(Pepperoni);
    }

    public sealed record Sausage : PizzaToppingKind
    {
        public static Sausage Instance { get; } = new();

        public override string ToString() => nameof(Sausage);
    }
}

public abstract record PizzaToppingAmount
{
    public sealed record Light : PizzaToppingAmount
    {
        public static Light Instance { get; } = new();

        public override string ToString() => nameof(Light);
    }

    public sealed record Normal : PizzaToppingAmount
    {
        public static Normal Instance { get; } = new();

        public override string ToString() => nameof(Normal);
    }

    public sealed record Extra : PizzaToppingAmount
    {
        public static Extra Instance { get; } = new();

        public override string ToString() => nameof(Extra);
    }
}

public abstract record PizzaSize
{
    public sealed record Small : PizzaSize
    {
        public static Small Instance { get; } = new();

        public override string ToString() => nameof(Small);
    }

    public sealed record Medium : PizzaSize
    {
        public static Medium Instance { get; } = new();

        public override string ToString() => nameof(Medium);
    }

    public sealed record Large : PizzaSize
    {
        public static Large Instance { get; } = new();

        public override string ToString() => nameof(Large);
    }
}

public sealed record Pizza
{
    public required PizzaSize Size { get; init; }
    public required HashMap<PizzaToppingKind, PizzaToppingAmount> Toppings { get; init; }
}

File: src/common/Pizza.fs
namespace common

open System

type ToppingKind =
    | Cheese
    | Pepperoni
    | Sausage

    static member toString value =
        match value with
        | Cheese -> nameof Cheese
        | Pepperoni -> nameof Pepperoni
        | Sausage -> nameof Sausage

    static member fromString value =
        match value with
        | _ when String.Equals(value, nameof Cheese, StringComparison.OrdinalIgnoreCase) -> Cheese
        | _ when String.Equals(value, nameof Pepperoni, StringComparison.OrdinalIgnoreCase) -> Pepperoni
        | _ when String.Equals(value, nameof Sausage, StringComparison.OrdinalIgnoreCase) -> Sausage
        | _ -> invalidOp ($"'{value}' is not a valid topping kind.")

type ToppingAmount =
    | Light
    | Normal
    | Extra

    static member toString value =
        match value with
        | Light -> nameof Light
        | Normal -> nameof Normal
        | Extra -> nameof Extra

    static member fromString value =
        match value with
        | _ when String.Equals(value, nameof Light, StringComparison.OrdinalIgnoreCase) -> Light
        | _ when String.Equals(value, nameof Normal, StringComparison.OrdinalIgnoreCase) -> Normal
        | _ when String.Equals(value, nameof Extra, StringComparison.OrdinalIgnoreCase) -> Extra
        | _ -> invalidOp ($"'{value}' is not a valid topping amount.")

type PizzaSize =
    | Small
    | Medium
    | Large

    static member toString value =
        match value with
        | Small -> nameof Small
        | Medium -> nameof Medium
        | Large -> nameof Large

    static member fromString value =
        match value with
        | _ when String.Equals(value, nameof Small, StringComparison.OrdinalIgnoreCase) -> Small
        | _ when String.Equals(value, nameof Medium, StringComparison.OrdinalIgnoreCase) -> Medium
        | _ when String.Equals(value, nameof Large, StringComparison.OrdinalIgnoreCase) -> Large
        | _ -> invalidOp ($"'{value}' is not a valid pizza size.")

type Pizza =
    { Size: PizzaSize
      Toppings: Map<ToppingKind, ToppingAmount> }
